### PR TITLE
feat(route): 增加知网作者最新发布的期刊文献

### PR DESCRIPTION
**Pull request by harveyqiu**
## 该 PR 相关 Issue / Involved issue
无

## 完整路由地址 / Example for the proposed route(s)

```routes
/cnki/author/000042423923
/cnki/author/000040489945
```

## 新 RSS 检查列表 / New RSS Script Checklist
  
- [x] 新的路由 New Route
  - [x] 跟随 [v2 路由规范](https://docs.rsshub.app/joinus/script-standard.html) Follows [v2 Script Standard](https://docs.rsshub.app/en/joinus/script-standard.html)
- [x] 文档说明 Documentation
  - [x] 中文文档 CN
  - [ ] 英文文档 EN
- [x] 全文获取 fulltext
  - [x] 使用缓存 Use Cache
- [ ] 反爬/频率限制 anti-bot or rate limit?
  - [ ] 如果有, 是否有对应的措施? If yes, do your code reflect this sign?
- [ ] [日期和时间](https://docs.rsshub.app/joinus/pub-date.html) [date and time](https://docs.rsshub.app/en/joinus/pub-date.html)
  - [ ] 可以解析 Parsed
  - [ ] 时区调整 Correct TimeZone
- [ ] 添加了新的包 New package added
- [ ] `Puppeteer`

## 说明 / Note
@alienmao 是我之前的github id。修改了之前的maintainer信息以及时获得维护提醒。

### CODE REVIEW

File: lib/v2/cnki/author.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+                Referer: authorPageUrl,
+            },
+        })
+        .then((res) => {
+            const $ = cheerio.load(res.data);
+            const list = $('.bignum')
+                .children()
+                .map((_, e) => ({
+                    title: $(e).find('a').first().text().trim(),
+                    link: 'https://chn.oversea.cnki.net/' + $(e).find('a').first().attr('href'),
```
Comment (TonyRL) [inline lib/v2/cnki/author.js:32]:
```suggestion
                    link: 'https://chn.oversea.cnki.net' + $(e).find('a').first().attr('href'),
```
Double slash

File: lib/v2/cnki/author.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+                }))
+                .get();
+            return list;
+        });
+
+    const items = await Promise.all(list.map((item) => ctx.cache.tryGet(item.link, () => utils.ProcessItem(item))));
+
+    ctx.state.data = {
+        title: `知网 ${authorName} ${companyName}`,
+        link: url,
```
Comment (TonyRL) [inline lib/v2/cnki/author.js:42]:
```suggestion
        link: authorPageUrl,
```

File: lib/v2/cnki/radar.js [hunk 3]
```diff
@@ -14,6 +14,12 @@ module.exports = {
                 source: ['/knavi/journals/:name/detail'],
                 target: '/cnki/journals/debut/:name',
             },
+            {
```
Comment (TonyRL) [inline lib/v2/cnki/radar.js:17]:
```js
        kns: [
            {
                title: '作者期刊文献',
                docs: 'https://docs.rsshub.app/journal.html#zhong-guo-zhi-wang-zuo-zhe-qi-kan-wen-xian',
                source: ['/kcms/detail/knetsearch.aspx', '/'],
                target: (_, url) => `/cnki/author/${new URL(url).searchParams.get('code')}`,
            },
        ],

File: lib/v2/cnki/author.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+const rootUrl = 'https://chn.oversea.cnki.net/kcms/detail/frame/knetlist.aspx?infotype=4&codetype=1&vl=&';
+const authorPageRootUrl = 'https://kns.cnki.net/kcms/detail/knetsearch.aspx?sfield=au&';
+
+module.exports = async (ctx) => {
+    const { code } = ctx.params;
+
+    const authorPageUrl = `${authorPageRootUrl}code=${code}`;
+
+    const res = await got.get(authorPageUrl);
```
Comment (TonyRL) [inline lib/v2/cnki/author.js:13]:
```suggestion
    const res = await got(authorPageUrl);
```

File: lib/v2/cnki/utils.js [hunk 5]
```diff
@@ -0,0 +1,26 @@
+const got = require('@/utils/got');
+const cheerio = require('cheerio');
+const { art } = require('@/utils/render');
+const path = require('path');
+
+const ProcessItem = async (item) => {
+    const detailResponse = await got.get(item.link);
```
Comment (TonyRL) [inline lib/v2/cnki/utils.js:7]:
```suggestion
    const detailResponse = await got(item.link);
```

**Comment by harveyqiu** [comment]
The url of author page has changed.
The original url is like https://kns.cnki.net/kcms/detail/knetsearch.aspx?sfield=au&code=000042423923.

Now the url has changed to https://kns.cnki.net/kcms2/author/detail?v=3_oUXeZjdDw0ZrbRwePiTE0B9IpzazwDg2M-gCtfij5H_i2Pdww5se8-IMGMgYieRJ6_gzZWlOHjFb8lNPsr3rXOKYoYIsyoMmcoEvjlxN4=&uniplatform=NZKPT. So you may be not able to get the code.


Turn on the developer tool of your browser. Then you get the "au-code". Just replace the code in ${rootUrl}code=${code} with "au-code". It should be ok.

image

If you have any further question, feel free to reply me.
@Derekmini

**Comment by Derekmini** [comment]
I apologize, I did not input the complete link in my previous message. The link I intended to ask about is the one that starts with `chn.oversea.cnki.net`: https://chn.oversea.cnki.net/kcms/detail/frame/knetlist.aspx?infotype=4&codetype=1&vl=&code=000042423923. May I ask where you found this link? Or throuth packet capture software?


Based on your code to `got` the paper list through this link, I discovered that the content returned not contain the paper list. This leads me to believe that the link has been updated or replaced. As I am unsure how you initially discovered this link(I still have not been able to find this link in packet capture software, even when accessing oversea.cnki through a proxy), I would like to ask for your advice on this matter.


Your code for obtaining the article list through this link is shown below, which I have followed:

https://github.com/DIYgod/RSSHub/blob/98cfccb65d8202e01d20937cef533d43fbb5a4fa/lib/v2/cnki/author.js#L20-L25


I have obtained the author's name and institutional affiliation by using Node in VS Code, but I am unable to `got` the paper list by the link which starts with `chn.oversea.cnki.net`.

**Comment by harveyqiu** [comment]
The implementation is inspired by https://github.com/DIYgod/RSSHub/commit/63e78432be53dc375cdb39206d77aea7510d96c4.

CNKI changes its website quite often. I just checked https://rsshub.app/cnki/author/000042423923. It does not work as expected. While I hope to fix this problem when I am free, I must apologize in advance as I am currently occupied serving as a data compliance lawyer and may not have the time available to address the issue at this moment. 

Feel free to fix the code if you want.

**Comment by Derekmini** [comment]
谢谢您的解答，我也是突然发现RSS阅读器好久没更新CNKI上的条目了，才尝试修改一下。

那我尝试一下有没有其他方式获取到最新的文章列表链接，如果有好的进展我会提请new pr！

再次感谢您的回复！很高兴与您交流！

**Comment by harveyqiu** [comment]
没事，到时候开PR也可以@我。

**Comment by Derekmini** [comment]
好的，ok👌

**Comment by harveyqiu** [comment]
我目前看了一下。
应该是以下中的一个：

- https://kns.cnki.net/restapi/knowledge-api/v1/experts/relations/resources?v=3uoqIhG8C443IK8pERLhi-uPd6xnhDw6bO0mo6lBCtrF5oSs5k0GkUhPujM_eLjF3INRvCa_F9p_XkFaROtojTHIV3LhbgsGtJol3MBsU7wT8sKk3kr6EwWry2f1qRV6&resource=SCDB&sequence=CF&size=10&sort=desc&start=1
- https://kns.cnki.net/restapi/knowledge-api/v1/experts/relations/resources?v=3uoqIhG8C443IK8pERLhi-uPd6xnhDw6bO0mo6lBCtrF5oSs5k0GkUhPujM_eLjF3INRvCa_F9p_XkFaROtojTHIV3LhbgsGtJol3MBsU7wT8sKk3kr6EwWry2f1qRV6&resource=SCDB&sequence=DFR&size=10&sort=desc&start=1
- https://kns.cnki.net/restapi/knowledge-api/v1/experts/relations/resources?v=3uoqIhG8C443IK8pERLhi-uPd6xnhDw6bO0mo6lBCtrF5oSs5k0GkUhPujM_eLjF3INRvCa_F9p_XkFaROtojTHIV3LhbgsGtJol3MBsU7wT8sKk3kr6EwWry2f1qRV6&sequence=PT&size=10&sort=desc&start=1&resource=CJFD

你在开发者工具的网络标签页，搜索knowledge-api应该就能搜到。
@Derekmini

**Comment by Derekmini** [comment]
原来是这样获取到的链接！ 那我去尝试`got`一下。

**Comment by Derekmini** [comment]
现在已经可以输出条目结果了


我目前只是在vs code中使用node实现，还未向RSSHub中改，代码如下：

```javascript
import got from 'got';
import cheerio from 'cheerio';

(async () => {
  try {
    const rootUrl = 'https://kns.cnki.net';
    const code = '3uoqIhG8C443IK8pERLhi-uPd6xnhDw6bO0mo6lBCtrF5oSs5k0GkUhPujM_eLjF3INRvCa_F9p_XkFaROtojTHIV3LhbgsGtJol3MBsU7wT8sKk3kr6EwWry2f1qRV6';
    // const authorPageUrl = `https://kns.cnki.net/kcms2/author/detail?v=${code}&uniplatform=NZKPT`;
    const authorPageUrl = `${rootUrl}/kcms2/author/detail?v=${code}&uniplatform=NZKPT`;
    
    const response = await got(authorPageUrl);
    const $ = cheerio.load(response.body);

    const authorName = $('#au-name').attr('value'); // 作者姓名

    console.log(authorName);

    // const url =`https://kns.cnki.net/restapi/knowledge-api/v1/experts/relations/resources?v=${code}&sequence=PT&size=10&sort=desc&start=1&resource=CJFD`;
    const url = `${rootUrl}/restapi/knowledge-api/v1/experts/relations/resources?v=${code}&sequence=PT&size=10&sort=desc&start=1&resource=CJFD`

    const response2 = await got(url, {headers: {Referer: authorPageUrl}});

    const jsonObj = JSON.parse(response2.body);
    const data = jsonObj.data.data;

    // console.log(data[0]);

    // 遍历所有对象并访问它们的属性
    for (let i = 0; i < data.length; i++) {
        // console.log(data[i].metadata);
        const metadata = data[i].metadata;
        const relations = data[i].relations;
        let title, date, author;
        for (let j = 0; j < metadata.length; j++) {
            const {name, value} = metadata[j];
            if (name === 'TI') {
                title = value;
            } else if (name === 'PT') {
                date = value;
            } else if (name === 'AU') {
                author = value;
            }
        }
        console.log('Title:', title);
        console.log('Date:', date);
        console.log('Author:', author);
        console.log('Link:', relations[0].url);
    }

  } catch (error) {
    console.log(error.response.body);
  }
})();
```

**Comment by harveyqiu** [comment]
> 原来是这样获取到的链接！ 那我去尝试`got`一下。

cnki的api变化的挺频繁的。我是重新看了一下获取论文列表的方式。之前的方式可能已经不适用了。

**Comment by Derekmini** [comment]
嗯嗯，对，现在去获取文章列表返回的是JSON格式的内容了，目前获取的内容变少了，没有了机构与摘要，现在只能看看文章题目了。

**Comment by Derekmini** [comment]
我在自己的库中提交了2个commit ，已经在RSSHub中实现了提及的基本功能，并且可以在RSS客户端正确查看获取到的文章。

已经将您发送Collaborators[申请](https://github.com/Derekmini/RSSHub/invitations)，希望您能同意，待您同意后我计划提交pr。

**Comment by harveyqiu** [comment]
> 我在自己的库中提交了2个commit ，已经在RSSHub中实现了提及的基本功能，并且可以在RSS客户端正确查看获取到的文章。
> 
> 已经将您发送Collaborators[申请](https://github.com/Derekmini/RSSHub/invitations)，希望您能同意，待您同意后我计划提交pr。
> 
> 因为我也好久没写路由了，而且我发现现在路由规范更加严格了，所以还是想与您共同完成该路由的修正。
> 
> 最终实现的流程为：访问旧的作者主页，获取作者名👉使用作者名+code作为参数，获取重定向的链接及新的code2👉访问knowledge-api获取条目
> 
> @harveyqiu

更改了部分代码，实现了获取摘要。
@Derekmini 
我认为可以提交开PR了。

**Comment by harveyqiu** [comment]
若你觉得可以，可以把你自己改为该路由的维护者。@Derekmini
抱歉个人工作较为麻烦，可能除了法定节假日之外的时间无法及时响应需求。

**Comment by Derekmini** [comment]
@harveyqiu 您的commit已经收到，感谢您添加获取摘要的功能，这样就完整了！

我看您把author更改为authorName，我原来的考虑是：有的文章还会有其他作者，原来的这部分代码会显示全部作者。我是在尝试其他作者code时发现的，只是咱们讨论的这个作者所有文章只有他自己的名字。

关于维护者，我想我把名字放在您后面比较妥当。我也是因为您快速的回复与高效的建议才能尽早摸清思路。

**Pull request merged**